### DISCUSSION TITLE
Add/extend incoming (registering) user with custom value (like hashed_password) without schemas usage.

### DISCUSSION THREAD

**Discussion by david-shiko** [fastapi-users/fastapi-users#587] [Q&A]
I want to add some custom values to every registering user (for example token, datetime, verification code) how I can do it without intervening in the source code?

**Comment by frankie567** [comment]
You can simply add the properties you wish to your `User` classes: https://frankie567.github.io/fastapi-users/configuration/model/#define-your-models
The API will automatically accept those new fields (actually, that's the reason why you need to define them and pass them as argument of `FastAPIUsers`).

For example:

```py
class User(models.BaseUser):
    first_name: Optional[str]


class UserCreate(models.BaseUserCreate):
    first_name: str


class UserUpdate(User, models.BaseUserUpdate):
    first_name: Optional[str]


class UserDB(User, models.BaseUserDB):
    first_name: str
```

I recommend you to check this links to better understand the concepts:

* https://fastapi.tiangolo.com/tutorial/extra-models/
* https://pydantic-docs.helpmanual.io/usage/models/

**Reply by david-shiko** [reply]
Thanks for a quick reply and sorry for an unclear question, but I mean that I want to implicity add some values to user (which are coming not from user as hashed_password for example or user_id).

In my case this is sms verification code that I want to assign for every registered user. So, psoted example are not my case.

**Reply by frankie567** [reply]
Ah, I see!

What you can do is to use [Pydantic validators](https://pydantic-docs.helpmanual.io/usage/validators/#validate-always) to generate the value automatically. Here is an example to add the creation date:

```py
from datetime import datetime, timezone

from pydantic import validator


class User(BaseUser):
    created_at: datetime = None

    @validator("created_at", pre=True, always=True)
    def default_created_at(cls, v):
        return v or datetime.now(timezone.utc)
```

The effect of `pre` and `always` is that the method will be called every time prior each other validation. If there is no value, a default is asssigned.

If you need to access other fields to generate yours, you can use [root validators](https://pydantic-docs.helpmanual.io/usage/validators/#root-validators) which let you access to the whole values.

**Reply by david-shiko** [reply]
I knowed about such option, but it's a bad approach.
I tried to do it in my old project and at some moment figured out that a code become a very unclear
(For example you should duplicate your logic (services.py file) into schemas file)
As i got from another question, pydantic not assumed to assign values (despite it can)
See https://github.com/samuelcolvin/pydantic/issues/2310#issuecomment-773920619

MAybe there are another way to do it?

**Reply by frankie567** [reply]
Not that I know of 🤷‍♂️

I don't see what's bad with having validators. If the logic you execute inside those is heavy, you could very well put it in another module and import it:

```py
from services.complicated_module import complicated_function


class User(BaseUser):
    complicated_value: str

    @validator("complicated_value", pre=True, always=True)
    def default_complicated_value(cls, v):
        return v or complicated_function()
```

If you use this validation logic several times across your models, you could also put it in a mixin class and inherit from it:

```py
class ComplicatedValueMixin(BaseModel):
    complicated_value: str

    @validator("complicated_value", pre=True, always=True)
    def default_complicated_value(cls, v):
        return v or complicated_function()


class User(BaseUser, ComplicatedValueMixin):
    pass


class Item(ComplicatedValueMixin):
    pass
```

**Reply by PuchatekwSzortach** [reply]
@frankie567 
Validators are for validation, not assigning data. Using them to assign data is breaking query command separation principle, and basically creating hard to follow code.

Also validators can't accept external configuration data without doing closures/lambdas/other tricks that again lead to hard to follow code.

Say that for whatever reason for every user registration request I want to save on new user instance field `label` that is computed based on some app configuration settings value (that varies between e.g. production and staging environments) and number of currently registered users/whatever else that comes from external data source.
Coercing pydantic validators to do that doesn't sound right.

As for a better solution - wouldn't subclassing `UserDatabase` instance (e.g. `fastapi_users.db.sqlalchemy.SQLAlchemyUserDatabase` if using `sqlalchemy`) one configures `fastapi_users.FastAPIUsers` instance with, and overwriting its `_make_user(...)` function be a valid solution for this problem? 
Constructor of customized `UserDatabase` instance could accept any data necessary to make required computations (e.g. settings instance, database connection string/sessions factory, etc)

**Reply by frankie567** [reply]
Why not, that could be a solution indeed.

What I usually do myself when I have heavy computations to do and store on the user is to use the `on_after_register` handler to send a task in a worker queue. This way, you could do things that take even minutes to do without blocking the user waiting for its response.

To further expand on my first proposal, there is now a better way than validators to assign default value thanks to a function: [`default_factory`](https://pydantic-docs.helpmanual.io/usage/models/#field-with-dynamic-default-value).
